restaking_program/src/ncn_delegate_token_account.rs

```diff
@@ -0,0 +1,87 @@
+use jito_bytemuck::AccountDeserialize;
+use jito_jsm_core::loader::{
+    load_signer, load_token_2022_program, load_token_account, load_token_mint, load_token_program,
+};
+use jito_restaking_core::ncn::Ncn;
+use solana_program::{
+    account_info::AccountInfo, entrypoint::ProgramResult, msg, program::invoke_signed,
+    program_error::ProgramError, pubkey::Pubkey,
+};
+
+/// Processes the ncn delegate token account instruction: [`crate::RestakingInstruction::NcnDelegateTokenAccount`]
+///
+/// This instruction handles the delegation of tokens from a token account managed by an NCN to a delegate account.
+/// The NCN admin might call this instruction to delegate authority over tokens managed by the NCN to another account.
+///
+/// # Arguments
+/// * `program_id` - The public key of the program, used to ensure the correct program is being executed.
+/// * `accounts` - A slice of `AccountInfo` representing the accounts required for this instruction.
+/// * `amount` - The number of tokens to delegate to the delegate account.
+///
+/// # Returns
+/// * `ProgramResult` - Returns `Ok(())` if the delegation is successful, otherwise returns an appropriate `ProgramError`.
+pub fn process_ncn_delegate_token_account(
+    program_id: &Pubkey,
+    accounts: &[AccountInfo],
+    amount: u64,
+) -> ProgramResult {
+    let [ncn_info, admin, token_mint, token_account, delegate, token_program_info] = accounts
+    else {
+        return Err(ProgramError::NotEnoughAccountKeys);
+    };
+
+    Ncn::load(program_id, ncn_info, false)?;
+    load_signer(admin, false)?;
+    load_token_mint(token_mint)?;
+    load_token_account(token_account, token_mint.key, token_program_info)?;
+
+    match (*token_mint.owner, *token_account.owner) {
+        (spl_token::ID, spl_token::ID) => {
+            load_token_program(token_program_info)?;
+        }
+        (spl_token_2022::ID, spl_token_2022::ID) => {
+            load_token_2022_program(token_program_info)?;
+        }
+        _ => {
+            msg!("token_mint and token_account owner does not match");
+            return Err(ProgramError::InvalidAccountData);
+        }
+    }
+
+    let ncn_data = ncn_info.data.borrow();
+    let ncn = Ncn::try_from_slice_unchecked(&ncn_data)?;
+
+    // The Ncn admin shall be the signer of the transaction
+    ncn.check_admin(admin)?;
```

Comment:
perhaps change withdraw_admin to delegate_admin here + operator + add check_delegate_admin

Reply:
we can likely remove the fee wallet in both NCN + operator - worth keeping the space in there for now so we don't break SDKs on testnet so you can do something like reserved_1: [u8; 32]

Reply:
I renamed `withdraw_admin` to `delegate_admin` and removed `withdraw_fee_wallet`.
Inside `delegate_token_account` ix, check the delegate_admin is signed and delegate_admin is correct.